### ISSUE TITLE
[Bug]: Linux: Double actions when handling app-command handlers: browser-backward and browser-forward #18322 

### ISSUE THREAD

**Issue by LunNova** [electron/electron#32763]
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

Confirmed to impact these versions:

**v16, v17.0.0, v20.1.0, v30.5.1, v31.6.0, v32.1.2**

Probably impacts all versions in between.

### What operating system are you using?

* Linux (NixOS, Ubuntu) - [testcase gist](https://gist.github.com/LunNova/a4c61b7082d17b8aa637217f763348bf) results in handling of app commands on both button down and button up
* Window - [testcase gist](https://gist.github.com/LunNova/a4c61b7082d17b8aa637217f763348bf) works as expected moving back/forward once per press

### Operating System Version

 Linux lun-hisame-nixos 5.15.19-xanmod1-tt #1-NixOS SMP PREEMPT Tue Jan 1 00:00:00 UTC 1980 x86_64 GNU/Linux  
Linux lun-hisame-nixos 6.11.0 #1-NixOS SMP PREEMPT_DYNAMIC Sun Sep 15 14:57:56 UTC 2024 x86_64 GNU/Linux  
Multiple distros and downstream apps relying on Electron reported in the comments.

### What arch are you using?

x64, aarch64

### Last Known Working Electron version

4

### Expected Behavior

When pressing back/forward mouse buttons on linux, either app command events should be created OR they should be automatically handled and cause navigation.

Two suggested approaches:

- To match Windows behaviour, we should emit app-command events and not cause navigation.
- To minimise breakage of existing Linux apps that don't handle app-command, app-command events should not be emitted and the navigation should still be done automatically

### Actual Behavior

When the mouse side buttons are used, app command events are created AND navigation is automatically done so if an app handles the app-command events per the documentation then navigation will happen twice.

https://github.com/electron/electron/pull/15441 added support for app-command handlers: browser-backward and browser-forward on Linux.

The docs describe how to make use of these here:
https://github.com/electron/electron/blob/v5.0.1/docs/api/browser-window.md#event-app-command-windows-linux

Our code is very similar here, and has been in place to support windows for years:
https://github.com/irccloud/irccloud-desktop/blob/master/app/main.js#L293-L311

When testing this on Ubuntu 19.04, this results in back and forward actions being triggered twice when using the appropriate mouse buttons.

### Testcase Gist URL

https://gist.github.com/LunNova/a4c61b7082d17b8aa637217f763348bf

### Additional Information

A previous ticket was closed #18322 but this is still a problem. I have asked for it to be reopened, but noone seems to have seen that.

Related third party ticket https://support.discord.com/hc/en-us/community/posts/1500000622702-Thumb-mouse-button-on-Linux-registers-twice

**Comment by mlaurencin** [comment]
Thanks for reporting this and helping to make Electron better!

Would it be possible for you to make a standalone testcase with only the code necessary to reproduce the issue? For example, [Electron Fiddle](https://www.electronjs.org/fiddle) is a great tool for making small test cases and makes it easy to publish your test case to a [gist](https://gist.github.com) that Electron maintainers can use. If you take the code from either of the examples you provided and the issue replicable in the Fiddle Gist, that will helpful. 

Stand-alone test cases make fixing issues go more smoothly: it ensure everyone's looking at the same issue, it removes all unnecessary variables from the equation, and it can also provide the basis for automated regression tests.

I'm adding the `blocked/need-repro` label for this reason. After you make a test case, please link to it in a follow-up comment.

**Comment by LunNova** [comment]
@mlaurencin

index.js
```
const {app, BrowserWindow} = require('electron');

app.on('window-all-closed', () => {
  if (process.platform !== 'darwin') app.quit()
})
app.whenReady().then(async () => {
  let mainWindow = new BrowserWindow({height: 600, width: 600});
  // Handle app command per docs except also refresh and forward
  // https://www.electronjs.org/docs/latest/api/browser-window/#event-app-command-windows-linux
  mainWindow.on('app-command', function (e, cmd) {
    switch (cmd) {
    case 'browser-backward':
      if (mainWindow.webContents.canGoBack()) {
        mainWindow.webContents.goBack();
      }
      break;
    case 'browser-forward':
      if (mainWindow.webContents.canGoForward()) {
        mainWindow.webContents.goForward();
      }
      break;
    case 'browser-refresh':
      mainWindow.webContents.reloadIgnoringCache();
      break;
    default:
      break;
    }
  });
  for (var i = 0; i <= 10; i++) {
    await mainWindow.loadURL("data:text/plain;," + i + "/10");
  }
});
```

On Windows: back/forward button presses will go back/forward one page at a time  
On Linux: back/forward button presses will go back/forward two pages instead

**Comment by LunNova** [comment]
https://gist.github.com/LunNova/a4c61b7082d17b8aa637217f763348bf

**Comment by lmcarreiro** [comment]
> I'm experiencing this issue with Slack's snap on Ubuntu 22.04 as well.

Same here, very annoying bug!

If I install Slack from another source instead of snap, would we have the same issue?

**Comment by hjkatz** [comment]
No, I've tried both versions and it seems to be a problem with Slack's imported version of Electron. I also reached out to Slack support and they said they were aware of the problem, but gave no timeline for an intended fix.

**Comment by diniamo** [comment]
I have this issue on Discord as well.

**Comment by edoardopirovano** [comment]
Also experiencing this on Slack in Ubuntu 22.04.

**Comment by rneatherway** [comment]
Same here. Is there any way that we could move the discussion forward on which solution might be workable?

> To match Windows behaviour, we should emit app-command events and not cause navigation.
> To minimise breakage of existing Linux apps that don't handle app-command, app-command events should not be emitted and the navigation should still be done automatically

@MarshallOfSound might you have an opinion? I ask as you reviewed https://github.com/electron/electron/pull/15441 and your profile indicates that you work at Slack right now.

**Comment by JFGHT** [comment]
Bump on this.

**Comment by LunNova** [comment]
Tested again on electron `v20.1.0`, still reproduces. Can we get the labels updated @mlaurencin ?

**Comment by AddisonG** [comment]
Still getting this in Slack (v4.28.171) on Ubuntu 20.04, and have had this for a few years now.

**Comment by mennoxx** [comment]
> No, I've tried both versions and it seems to be a problem with Slack's imported version of Electron. I also reached out to Slack support and they said they were aware of the problem, but gave no timeline for an intended fix.

Thats what they already said, when i reported the issue to Slack 2 years ago unfortunately.. I don't expect Slack to fix this anymore while it still annoyes me every single day.

**Comment by wbollock** [comment]
bump 🙏

**Comment by LunNova** [comment]
It's still broken.

**Comment by lmcarreiro** [comment]
Very annoying issue. I'm getting on 4.29.149 64-bit Ubuntu 22.04

**Comment by TomerGodinger** [comment]
Yes, still happening and immensely frustrating.

**Comment by naclonts** [comment]
Happening for me on Slack Debian 11, as well.

**Comment by jason-h-35** [comment]
bump

**Comment by kznrluk** [comment]
bump, Slack desktop on Arch linux and XFCE4.

**Comment by IrrerPolterer** [comment]
Still experiencing the issue in Slack.
Mouse-Back and Mouse-Forward each register twice (on button press and release)

**Comment by weirdcease** [comment]
Experiencing this issue with WebCatalog on Linux.

**Comment by BinaryVixen899** [comment]
Still experienceing this with the Slack app in Ubuntu `23.04`

**Comment by jonathan-saindon** [comment]
I am experiencing this as well on `Mint 21.1`. Currently using Slack version `4.32.127 64-bit`

**Comment by mpearson** [comment]
Seeing this bug on Ubuntu 18.04 as well as 22.04. Slack version 4.32.127 64-bit. Extremely annoying. I have several other Electron apps which do not have this problem, e.g. VS Code and Discord, so presumably it's just an electron version issue?

**Comment by ReillyBrogan** [comment]
> Seeing this bug on Ubuntu 18.04 as well as 22.04. Slack version 4.32.127 64-bit. Extremely annoying. I have several other Electron apps which do not have this problem, e.g. VS Code and Discord, so presumably it's just an electron version issue?

Most likely the apps that do not have this issue are explicitly ignoring either the app-command events or the navigation events.

**Comment by gorlak** [comment]
bump

**Comment by wugaosheng123** [comment]
hello @codebytere 
I think this is a problem with the example code in the docs. `browser-backward` and ` browser-forward ` events are handled by default to navigate the page, so  shouldn't have to perform `win.webContents.goBack()` in `win.on('app-command', (e, cmd)......`, or else clicking the mouse button once navigates the page twice.
![图片](https://github.com/electron/electron/assets/88312105/3262f2df-e20a-4ba4-92b5-75d2f8bcf9c9)

But I didn't find the default handing of `browser-backward` and ` browser-forward `  in electron's code (which performs the navigation of the page), and I'm guessing this is done by chromium.If so, is this  [issure](https://github.com/electron/electron/pull/15441) meaningless？Because `browser-backward` and ` browser-forward ` already has a default action, unless he is told to perform another behavior instead of doing page navigation

**Comment by shawn-mccool-mollie** [comment]
Is there any work around? We have many people using slack and this is a significant usability issue.

**Comment by TaylorMichaelHall** [comment]
> Is there any work around? We have many people using slack and this is a significant usability issue.

The keyboard shortcuts work. Another option on Linux is to use the Slack webapp instead of the desktop app.

**Comment by lsaintier** [comment]
Today my compagny rollout a new slack release and it's still there. I can't believe it's still not fixed.

**Comment by alangalvino** [comment]
Still experiencing the issue in Slack.

**Comment by MKimiSH** [comment]
One workaround for Slack (works on Ubuntu 22.04) with an additional right click: 
1. Find a place where right click calls out a text editing dropdown from the system (Copy, Paste, Cut, etc.), not the one from Slack (Add reaction, reply in thread, etc.). I usually do it on the blank area to the right of the channel title. 
2. Right click in that region, and see the copy-paste dropdown. 
3. Move the cursor to be **still in Slack** but **outside the copy-paste dropdown**. This is important, because when we click forward/backward when the cursor is in the dropdown, nothing happens (at least for me). 
4. Click forward/backward. This should only move forward/backward by one page, not two.

**Comment by eizieizi** [comment]
itsdanomano found a temporary solution for the issue (posted on reddit), but it has to be re-done after every app update.
https://www.reddit.com/r/Slack/comments/qfodd7/when_i_press_back_on_a_mouse_in_the_slack_app_on/

Also for Snap its not so easy as the snap files are read only. Maybe one could work arround the issue with file mounts.

**Comment by vtushevskiyNV** [comment]
> itsdanomano found a temporary solution for the issue (posted on reddit), but it has to be re-done after every app update. https://www.reddit.com/r/Slack/comments/qfodd7/when_i_press_back_on_a_mouse_in_the_slack_app_on/
> 
> Also for Snap its not so easy as the snap files are read only. Maybe one could work arround the issue with file mounts.

doesn't work for me

**Comment by genka-pokos** [comment]
> > itsdanomano found a temporary solution for the issue (posted on reddit), but it has to be re-done after every app update. https://www.reddit.com/r/Slack/comments/qfodd7/when_i_press_back_on_a_mouse_in_the_slack_app_on/
> > Also for Snap its not so easy as the snap files are read only. Maybe one could work arround the issue with file mounts.
> 
> doesn't work for me

Try the command from the mythin's message. Works for me

**Comment by marcodiiga** [comment]
I also would love to see this fixed: I must reapply the reddit fix each time slack gets updated unfortunately.

**Comment by theodotdot** [comment]
Has anyone noticed the above fix not working anymore?

Stopped working for me running slack-desktop 4.37.101 on Ubuntu 20.04

**Comment by genka-pokos** [comment]
> Has anyone noticed the above fix not working anymore?
> 
> Stopped working for me running slack-desktop 4.37.101 on Ubuntu 20.04

Try this one by lalten https://www.reddit.com/r/Slack/comments/qfodd7/comment/kxo523r/, works for me

**Comment by theodotdot** [comment]
Ah that did fix it, I missed it in the reddit thread. So it seems like the above link is the fix that is working for the latest versions, thank you!

**Comment by tiaraju** [comment]
Still happening on ubuntu 22.04. Same issue for me.
Really frustrating 😢

**Comment by fnagua1** [comment]
Same issue here in Ubunto 22.04 and Slack 4.38.125.
Do we have an ETA for the fix? 

Thank you in advance!

**Comment by genka-pokos** [comment]
Since the last Slack update, the Reddit fix is no longer working. This is ridiculous, this issue is two years old, and all that needs to be done is to handle buttons only on release, not on push.

**Comment by eizieizi** [comment]
> Still happening on ubuntu 22.04. Same issue for me. Really frustrating 😢
> 
> it's the fix supposed to work by having slack installed by snap?

No, as SNAP images use their own filesystem, like a container this is more difficult to achieve. I switched from snap back to classic installed slack.

**Comment by AleksandrKls** [comment]
Same issue here in Ubuntu 24.04 and Slack 4.39.88

**Comment by mpearson** [comment]
Bump, still seeing this on Slack 4.39.95 (installed via Snap)  / Ubuntu 22.04.

**Comment by LunNova** [comment]
Ran the same testcase on Electron v30.5.1, v31.6.0, v32.1.2. Still broken, moves back once on button down and a second time on button release.

**Comment by eliliam** [comment]
Electron team, why can we not fix this?

**Comment by genka-pokos** [comment]
Any news on this?

**Comment by vtushevskiyNV** [comment]
bump

**Comment by Tejeev** [comment]
How is this still a thing?  It's frustrating af and I would really really appreciate it if we would finally get forward and backwards working, or at least post a workaround of some sort

**Comment by Tejeev** [comment]
@mlaurencin is there any hope here?

**Comment by Tejeev** [comment]
Looks like folks were updating their app.sar but I don't see one of those in the snap, or I don't know how snaps really work and kinda hate them... probably the later

**Comment by shinebayar-g** [comment]
Flatpak or Snap doesn't seem to provide .asar files, so I use native package managers and edit the files manually for now.

**Comment by Tejeev** [comment]
oh, like install from a .deb or something and then edit the file?  Snap is unfortuantely native in Ubuntu these days. I keep telling myself I'm going to change distros, but haven't had the time to migrate my configurations and etc to a new distro, let alone pick one.

**Comment by Tejeev** [comment]
Ok, so now I need to know how to handle this in the apps whcih do not have a .deb.  I've got one or two electron app based tools I use for work which are only published as snaps.

**Comment by frends-dani** [comment]
bump

**Comment by TomerGodinger** [comment]
My god, I restarted my computer today, two years after last writing here, and this has finally **stopped** happening in Slack for me.
(Ubuntu 22.04.5, Slack version 4.43.43 64-bit.)
I don't know what caused the change, but I'm just really happy to finally stop having this issue! 😁

**Comment by umpirsky** [comment]
I can confirm it finally stopped for me on `Ubuntu 24.04.2 LTS` too!

**Comment by AleksandrKls** [comment]
Wow, I can confirm too

`Ubuntu 24.04.2 LTS`

"Let the ticket breathe" and the problem solved itself

**Comment by eliliam** [comment]
Same! What was the thing that actually fixed this issue?

**Comment by amcsi** [comment]
Looks like it's fixed for me also on `Ubuntu 24.04.2 LTS`. Not sure if it was the distro or some newer Electron version.

**Comment by lythandas** [comment]
Working on `Ubuntu 24.04.2 LTS` confirmed! Great news!

**Comment by mlaurencin** [comment]
Thank you to everyone that has confirmed this issue is resolved for them using `Ubuntu 24.04.2 LTS`. Given the number of confirmations, I'm going to go ahead and close this issue. If you are still experiencing it, rather than comment here, please file a new bug report with all the relevant information.

**Comment by umpirsky** [comment]
Anyone knows what fixed it?
